rust/alluxiocommon/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+                            return Err(PyValueError::new_err(err.to_string()));
+                        },
+                    }
+                } else {
+                    return Ok(Self {
+                        max_threads: max_concurrency,
+                        ondemand_pool: true,
+                        thread_pool: None,
+                        request_client: None,
+                    });
+                }
+            },
+            None => {
+                return Ok(Self {
+                    max_threads: max_concurrency,
+                    ondemand_pool: true,
+                    thread_pool: None,
+                    request_client: None,
+                });
+            }
+        }
+    }
+
+    fn make_multi_http_req(self_: PyRef<'_, Self>, urls: Vec<String>) -> PyResult<PyObject> {
+        let num_reqs = urls.len();
+        let mut content_results = Vec::with_capacity(num_reqs);
+        let mut senders = Vec::with_capacity(num_reqs);
+        for _ in 0..num_reqs {
+            senders.push(None);
+        }
+
+        let thread_pool /*: Arc<ThreadPool>*/ = match self_.ondemand_pool {
+            true => {
+                match create_pool(cmp::min(self_.max_threads, num_reqs),
+                                  String::from(DEFAULT_THREADPOOL_NAME))
+                {
+                    Ok(pool) => Arc::new(pool),
+                    Err(err) => {
+                        PyException::new_err(err.to_string())
+                            .restore(self_.py());
+                        return Err(PyErr::fetch(self_.py()));
+                    },
+                }
+            },
+            false => {
+                Arc::clone(&(self_.thread_pool.as_ref().unwrap())) // it can't be None here once instantiated
+            }
+        };
+
+        let request_client = match self_.ondemand_pool {
+            true => {
+                Arc::new(Client::new())
+            },
+            false => {
+                Arc::clone(&(self_.request_client.as_ref().unwrap()))
+            }
+        };
+        for i in 0..num_reqs {
+            let url_owned = urls[i].to_owned();
+            let client_clone = Arc::clone(&(request_client));
+            let (send, recv) = tokio::sync::oneshot::channel();
+            let install_res = thread_pool.install(move || -> Result<(), reqwest::Error> {
+                let body = perform_http_get(url_owned.as_str(), client_clone.as_ref());
+                send.send(body).unwrap();
+                Ok(())
+            });
+            match install_res {
+                Ok(_success) => {},
+                Err(err) => {
+                    PyException::new_err(err.to_string())
+                        .restore(self_.py());
+                    return Err(PyErr::fetch(self_.py()));
+                },
+            }
+            senders[i] = Some(recv);
+        }
+
+        for sender in senders {
+            let result = sender.unwrap().blocking_recv();
```

Comment:
I feel like this could be simplified with rayon::iter::ParallelIterator.try_for_each. The output is ordered so in that case we can get rid of the oneshot channel as well.

Reply:
yeah u r right i will try that, i plan to clean up this lib after i check this extension it, will address this in later PR